### DISCUSSION TITLE
Matcher pattern with NOT_IN and extension attributes

### DISCUSSION THREAD

**Discussion by vito-detullio-finconsgroup** [explosion/spaCy#9789] [Help: Coding & Implementations] [answered]
I'm trying to write a `Matcher` pattern that use a pair of extension attributes that I set in my pipeline, but I have errors:

what I'm trying to achieve is a pattern that matches every sequence of tokens that is not an `end_phrase` *and* a `section_marker` (that I have implemented as extension attributes). So if a token is an `end_phrase`, but not a `section_marker` (or viceversa, or neither) should be accepted, but if is both of them, should stop the match.

I tried to write something like the following snippet:

```python
nlp = ... 
pattern = [{
	"OP": "+",
	"_": {
		"NOT_IN": [{
				"end_phrase": True
			},
			{
				"section_marker": True
			}
		]

	}
}]
matcher = Matcher(nlp.vocab, validate=True)
matcher.add('mymatch', [pattern], greedy='LONGEST')
```

As I have already used `NOT_IN` in other scenarios (with `POS` or `ENT_TYPE` values, mostly), but I have this error:

```
  File "spacy\matcher\matcher.pyx", line 128, in spacy.matcher.matcher.Matcher.add
  File "spacy\matcher\matcher.pyx", line 720, in spacy.matcher.matcher.init_pattern
TypeError: an integer is required
```

What is the "right" way to write the pattern?

**Comment by adrianeboyd** [comment, marked as answer]
The `_` values can only be specified individually, so you'd want something like:

```
{"_":` {"end_phrase": True, "section_marker": True}}
```

**Reply by vito-detullio-finconsgroup** [reply]
Maybe I'm missing something, but this is exactly the opposite of the pattern I want - as in: this pattern capture *only* tokens that are both with `end_phrase` and `section_marker` as `true`.
How can I use this?

**Reply by adrianeboyd** [reply]
You'd have to add multiple patterns for each combination that you want to match.

Edited to add: Or to keep a pattern with only one attribute, you could add a third custom extension with a `getter` that combines the values however you need for your pattern.
